pkg/apis/mesh/config_manager.go

```diff
@@ -0,0 +1,128 @@
+package mesh
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Keeps track of sidecar agent versions.
+// First key is a version (x.x.x), with a map of agents that run that version.
+type agentVersions map[string]map[string]struct{}
+
+// ConfigManager manages the in-memory mesh configuration.
+type ConfigManager struct {
+	AgentVersions agentVersions
+	meshConfig    FullMeshConfig
+	sync.RWMutex
+}
+
+// NewConfigManager creates a new mesh config manager.
+func NewConfigManager(config FullMeshConfig) *ConfigManager {
+	m := &ConfigManager{
+		AgentVersions: make(agentVersions),
+	}
+	m.SetConfig(config)
+
+	return m
+}
+
+// SetConfig sets the mesh config.
+func (m *ConfigManager) SetConfig(config FullMeshConfig) {
+	m.Lock()
+	defer m.Unlock()
+
+	m.meshConfig = config
+}
+
+// GetConfig returns mesh config.
+func (m *ConfigManager) GetConfig() FullMeshConfig {
+	m.Lock()
+	defer m.Unlock()
```

Comment:
I see this is a RWMutex, we should probably be using RLock/RUnlock for all reads.